src/algebra/group/hom.lean

```diff
@@ -184,15 +247,287 @@ attribute [instance] inv.is_group_hom neg.is_add_group_hom
+attribute [to_additive add_monoid_hom.rec_on] monoid_hom.rec_on
+attribute [to_additive add_monoid_hom.sizeof] monoid_hom.sizeof
+attribute [to_additive add_monoid_hom.to_fun] monoid_hom.to_fun
+
+namespace monoid_hom
+variables {M : Type*} {N : Type*} {P : Type*} [monoid M] [monoid N] [monoid P]
+variables {G : Type*} {H : Type*} [group G] [comm_group H]
+
+@[extensionality] def ext (f g : M →* N) (h : (f : M → N) = g) : f = g :=
+by cases f; cases g; cases h; refl
+
+/-- If f is a monoid homomorphism then f 1 = 1. -/
+@[simp] lemma map_one (f : M →* N) : f 1 = 1 := f.map_one'
+
+/-- If f is a monoid homomorphism then f (a * b) = f a * f b. -/
+@[simp] lemma map_mul (f : M →* N) (a b : M) : f (a * b) = f a * f b := f.map_mul' a b
+
+instance {M : Type*} {N : Type*} [monoid M] [monoid N] (f : M →* N) :
+  is_monoid_hom (f : M → N) :=
+{ map_mul := f.map_mul,
+  map_one := f.map_one }
+
+instance {G : Type*} {H : Type*} [group G] [group H] (f : G →* H) :
+  is_group_hom (f : G → H) :=
+{ map_mul := f.map_mul }
+
+/-- The identity map from a monoid to itself. -/
+def id (M : Type*) [monoid M] : M →* M :=
+{ to_fun := id,
+  map_one' := rfl,
+  map_mul' := λ _ _, rfl }
+
+/-- Composition of monoid morphisms is a monoid morphism. -/
+def comp (hnp : N →* P) (hmn : M →* N) : M →* P :=
+{ to_fun := hnp ∘ hmn,
+  map_one' := by simp,
+  map_mul' := by simp,
+}
+
+protected def one : M →* N :=
+{ to_fun := λ _, 1,
+  map_one' := rfl,
+  map_mul' := λ _ _, (one_mul 1).symm }
+
+instance : has_one (M →* N) := ⟨monoid_hom.one⟩
+
+/-- The product of two monoid morphisms is a monoid morphism if the target is commutative. -/
+protected def mul {M N} [monoid M] [comm_monoid N] (f g : M →* N) : M →* N :=
+{ to_fun := λ m, f m * g m,
+  map_one' := show f 1 * g 1 = 1, by simp,
+  map_mul' := begin intros, show f (x * y) * g (x * y) = f x * g x * (f y * g y),
+    rw [f.map_mul, g.map_mul, ←mul_assoc, ←mul_assoc, mul_right_comm (f x)], end
+}
+
+instance {M N} [monoid M] [comm_monoid N] : has_mul (M →* N) := ⟨monoid_hom.mul⟩
+
+/-- (M →* N) is a comm_monoid if N is commutative. -/
+instance {M N} [monoid M] [comm_monoid N] : comm_monoid (M →* N) :=
+{ mul := (*),
+  mul_assoc := by intros; ext; apply mul_assoc,
+  one := 1,
+  one_mul := by intros; ext; apply one_mul,
+  mul_one := by intros; ext; apply mul_one,
+  mul_comm := by intros; ext; apply mul_comm}
+
+/-- Group homomorphisms preserve inverse. -/
+@[simp] theorem map_inv {G H} [group G] [group H] (f : G →* H) (g : G) : f g⁻¹ = (f g)⁻¹ :=
+eq_inv_of_mul_eq_one $ by rw [←f.map_mul, inv_mul_self, f.map_one]
+
+/-- Group homomorphisms preserve division. -/
+@[simp] theorem map_div {G H} [group G] [group H] (f : G →* H) (g h : G) :
+  f (g * h⁻¹) = (f g) * (f h)⁻¹ := by rw [f.map_mul, f.map_inv]
+
+/-- Makes a group homomomorphism from a proof that the map preserves multiplication. -/
+def mk' {f : M → G} (map_mul : ∀ a b : M, f (a * b) = f a * f b) : M →* G :=
```

Comment:
I think it's just a bit confusing having `mk'` and `mk''`. I would keep `mk''` and get rid of `mk'`

Reply:
OK!